source/SH3/graphics/texture.cpp

```diff
@@ -9,52 +9,340 @@
 
     if(ret != sizeof(header))
-        die("sh3_texture::Load( ): Header only read %d bytes of data! It should have been %d", ret, sizeof(header)); // TODO: Fallback on dummy texture (like Source's purple and black checkboard)
+        die("sh3_texture::Load( ): ReadData( ) != sizeof(header)!");
 
-    if(header.bpp == 8)
-        Log(LogLevel::WARN, "WARNING: 8-bit textures are currently unsupported! They WILL look like nothing but pixel garbage.");
+    // Check for the pesky 64-byte A7A7A7A7 header that sometimes precedes our texture header
+    if(header.batchHeaderMarker == 0x00000000 && header.batchSize == 0xA7A7A7A7) // AHA!
+    {
+        offset = 0x40; // Skip the unknown header if it exists
 
-    Log(LogLevel::INFO, "Texture info:\n\ttexWidth:%d\n\ttexHeight:%d\n\ttexSize:%d\n\tbpp:%d", header.texWidth, header.texHeight, header.texSize, header.bpp);
+        file.Seek(offset, std::ios_base::beg);
+        file.ReadData(&header, sizeof(header), e);
 
-    file.Seek(sizeof(header) + header.dataOffset, std::ios_base::beg);
+        if(ret != sizeof(header))
+            die("sh3_texture::Load( ): ReadData( ) != sizeof(header)!");
+    }
 
-    data.resize(header.texSize);
+    if(header.texSize == (header.texWidth * header.texHeight* 4))
+    {
+        header.bpp = 32; // Thanks KONAMI!
+    }
 
-    if(header.texSize != header.texWidth * header.texHeight * header.bpp)
+    // Now that we're done that, we can check perform some sanity checks on our texture!
+    if(header.texSize != header.texWidth * header.texHeight * (header.bpp / 8))
     {
-        Log(LogLevel::ERROR, "sh3_texture::Load( ): texWidth * texHeight * bpp != header.texSize!");
-        return;
+        Log(LogLevel::WARN, "sh3_texture::Load( ): Warning, texSize != width * height * (bpp / 8)!");
+        return; // TODO: Bind a color shader here
     }
 
-    ret = file.ReadData(&data[0], header.texSize, e);
-    if(ret != header.texSize)
-        Log(LogLevel::WARN, "sh3_texture::Load( ): Only read %d out of %d bytes!", ret, header.texSize);
+    data.resize(header.texSize); // Early data resize (if it's an 8bpp texture, it will be resized anyway)
 
-    // Now actually create a logical texture with OpenGL on the gpu
-    glGenTextures(1, &tex);
+    if(header.bpp == PixelFormat::PALETTE)
+    {
+        palette_info         pal_header;
+        std::vector<bgra>    palette;    // Palette Data (I think this is BGRA)
 
-    glBindTexture(GL_TEXTURE_2D, tex);
-    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, header.texWidth, header.texHeight, 0, GL_BGRA, GL_UNSIGNED_BYTE, &data[0]);
-    glBindTexture(GL_TEXTURE_2D, 0); // Unbind this texture
+        // First, we need to seek to the palette and read it in.
+        file.Seek(offset + header.batchHeaderSize + header.texFileSize, std::ios_base::beg);
+        file.ReadData(&pal_header, sizeof(pal_header), e);
+
+        ASSERT(pal_header.bytes_per_pixel == sizeof(palette.front()));
```

Comment:
In most languages, `assert`s are features used for debugging, although it can usually be controlled via a separate flag. For the `assert` in the standard library, once has to define the `NDEBUG` macro to get rid of them, which CMake does for the release configuration.

As a side-note, since `assert`s are removed from release-builds, I'm not sure if the optimizer can make the same assumptions. We might want to modify `ASSERT` to do `if(x) __builtin_unreachable()` when disabled.